x-pack/plugins/security_solution/server/endpoint/routes/actions/file_upload_handler.ts

```diff
@@ -0,0 +1,136 @@
+  SecuritySolutionRequestHandlerContext,
+  HapiReadableStream,
+} from '../../../types';
+import type { EndpointAppContext } from '../../types';
+import { errorHandler } from '../error_handler';
+
+export const registerActionFileUploadRoute = (
+  router: SecuritySolutionPluginRouter,
+  endpointContext: EndpointAppContext
+) => {
+  if (!endpointContext.experimentalFeatures.responseActionUploadEnabled) {
+    return;
+  }
+
+  const logger = endpointContext.logFactory.get('uploadAction');
+
+  router.post(
+    {
+      path: UPLOAD_ROUTE,
+      validate: UploadActionRequestSchema,
+      options: {
+        authRequired: true,
+        tags: ['access:securitySolution'],
+        body: {
+          accepts: ['multipart/form-data'],
+          output: 'stream',
+          maxBytes: endpointContext.serverConfig.maxUploadResponseActionFileBytes,
+        },
+      },
+    },
+    withEndpointAuthz(
+      { all: ['canWriteFileOperations'] },
+      logger,
+      getActionFileUploadHandler(endpointContext)
+    )
+  );
+};
+
+export const getActionFileUploadHandler = (
+  endpointContext: EndpointAppContext
+): RequestHandler<never, never, UploadActionRequestBody, SecuritySolutionRequestHandlerContext> => {
+  const logger = endpointContext.logFactory.get('uploadAction');
+  const maxFileBytes = endpointContext.serverConfig.maxUploadResponseActionFileBytes;
+
+  return async (context, req, res) => {
+    const user = endpointContext.service.security?.authc.getCurrentUser(req);
+    const esClient = (await context.core).elasticsearch.client.asInternalUser;
+    const fileStream = req.body.file as HapiReadableStream;
+    const { file: _, parameters: userParams, ...actionPayload } = req.body;
+    const uploadParameters: ResponseActionUploadParameters = {
+      ...userParams,
+      file: {
+        file_id: '',
+        file_name: '',
+        sha256: '',
+        size: 0,
+      },
+    };
+
+    try {
+      const createdFile = await createFile({
+        esClient,
+        logger,
+        fileStream,
+        agents: actionPayload.endpoint_ids,
+        maxFileBytes,
+      });
+
+      uploadParameters.file.file_id = createdFile.file.id;
+      uploadParameters.file.file_name = createdFile.file.name;
+      uploadParameters.file.sha256 = createdFile.file.hash?.sha256;
+      uploadParameters.file.size = createdFile.file.size;
+    } catch (err) {
+      return errorHandler(logger, res, err);
+    }
+
+    try {
+      const casesClient = await endpointContext.service.getCasesClient(req);
+      const { action: actionId, ...data } = await endpointContext.service
+        .getActionCreateService()
+        .createAction<ResponseActionUploadOutputContent, ResponseActionUploadParameters>(
+          {
+            ...actionPayload,
+            parameters: uploadParameters,
+            command: 'upload',
+            user,
+          },
+          { casesClient }
+        );
+
+      await setFileActionId(esClient, logger, data);
```

Comment:
Q: What happens if the `setFileActionId` fails?

Reply:
Then we fail the API request.  But - thats a good point that I will need to think through. At this point in the code, a failure is unlikely, but perhaps if it happens to fail here, we just catch it and log it, but allow the api call to succeed, since the action has already been "sent" to the Endpoint.  Failing here as is right now will cause the file to be deleted (in the `catch()` below)` which would in turn cause the endpoint to fail to down the file. 

Good catch @ashokaditya . Will correct it in next PR.

Reply:
Thanks for the explanation. Sounds good with later updates. 🚀